nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/content/createindex/CreateIndexFacetImpl.java

```diff
@@ -0,0 +1,170 @@
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.repository.view.payloads.StringPayload;
+import org.sonatype.repository.helm.internal.content.HelmContentFacet;
+import org.sonatype.repository.helm.internal.createindex.CreateIndexFacet;
+import org.sonatype.repository.helm.internal.hosted.HelmHostedFacet;
+import org.sonatype.repository.helm.internal.metadata.ChartIndex;
+import org.sonatype.repository.helm.internal.orient.createindex.HelmIndexInvalidationEvent;
+import org.sonatype.repository.helm.internal.util.YamlParser;
+
+import com.google.common.eventbus.AllowConcurrentEvents;
+import com.google.common.eventbus.Subscribe;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport.State.STARTED;
+import static org.sonatype.repository.helm.internal.AssetKind.HELM_INDEX;
+
+/**
+ * Facet for rebuilding Helm index.yaml files
+ *
+ * @since 1.0.11
+ */
+@Named
+public class CreateIndexFacetImpl
+    extends FacetSupport
+    implements CreateIndexFacet, Asynchronous
+{
+  private final EventManager eventManager;
+
+  private final long interval;
+
+  private static final String INDEX_YAML = "index.yaml";
+
+  private static final String INDEX_YAML_CONTENT_TYPE = "text/x-yaml";
+
+  private final AtomicBoolean acceptingEvents = new AtomicBoolean(true);
+
+  private final static String API_VERSION = "v1";
+
+  private final YamlParser yamlParser;
+
+  //Prevents the same event from being fired multiple times
+  private final AtomicBoolean eventFired = new AtomicBoolean(false);
+
+  @Inject
+  public CreateIndexFacetImpl(final EventManager eventManager,
+                              final YamlParser yamlParser,
+                              @Named("${nexus.helm.createrepo.interval:-1000}") final long interval)
+  {
+    this.eventManager = checkNotNull(eventManager);
+    this.interval = interval;
+    this.yamlParser = checkNotNull(yamlParser);
+  }
+
+  @Subscribe
+  @Guarded(by = STARTED)
+  @AllowConcurrentEvents
+  public void on(RepositoryCreatedEvent createdEvent) {
+    // at repository creation time, create index.yaml with empty entries
+    log.debug("Initializing index.yaml for hosted repository {}", getRepository().getName());
+    invalidateIndex();
+  }
+
+  @Subscribe
+  public void on(final HelmIndexInvalidationEvent event) {
+    if (shouldProcess(event)) {
+      acceptingEvents.set(false);
+      maybeWait(event);
+
+      log.info("Rebuilding helm index for repository {}", getRepository().getName());
+
+      try {
+        acceptingEvents.set(true);
+        eventFired.set(false);
+
+        createIndexYaml();
+      }
+      finally {
+        log.info("Finished rebuilding helm index for repository {}", getRepository().getName());
+      }
+    }
+  }
+
+  //protected void updateIndexYaml(final TempBlob indexYaml) {
```

Comment:
Do we need this commented block?

Reply:
most likely we will need to return in once AssetEvent engine will be implemented. Ok let delete it for now 89ab4890de98a214b4e7bfa1b07d5aab872005d2